bootstrap/run.py

```diff
@@ -104,18 +104,87 @@ def run(path_opts=None):
         engine.train()
 
 
+def activate_debugger(): 
+    if Options()['misc'].get('debug', False):
+        Logger().set_level(Logger.DEBUG)
+        Logger()('Debug mode activated.', log_level=Logger.DEBUG)
+
+
+def activate_profiler():
+    if sys.version_info[0] == 3: #PY3
+        import builtins
+    else:
+        import __builtin__ as builtins
+    if Options()['misc'].get('profile', False):
+        # if profiler is activated, associate line_profiler
+        Logger()('Activating line_profiler...')
+        try:
+            import line_profiler
+        except ModuleNotFoundError:
+            Logger()('Failed to import line_profiler.', log_level=Logger.ERROR, raise_error=False)
+            Logger()('Please install it from https://github.com/rkern/line_profiler', log_level=Logger.ERROR, raise_error=False)
+            return
+        prof = line_profiler.LineProfiler()
+        builtins.__dict__['profile'] = prof
+    else:
+        # otherwise, create a blank profiler, to disable profiling code
+        builtins.__dict__['profile'] = lambda func: func
+        prof = None
+    return prof
+
+
+def process_profiler(prof):
+     if Options()['misc'].get('profile', False):
+        # unavoidable lazy import -- only if profiler is enabled
+        from line_profiler import show_text
+        results_file = os.path.join(Options()['exp']['dir'], 'profile_results.lprof')
+        Logger()('Saving profiler results to {}...'.format(results_file))
+        prof.dump_stats(results_file)
+        stats = prof.get_stats()
+        textio = io.StringIO()
+        show_text(stats.timings, stats.unit, stream=textio)
+        lines = textio.getvalue()
+        Logger()('Printing profiling results', log_level=Logger.SYSTEM)
+        for line in lines.splitlines():
+            Logger()(line, log_level=Logger.SYSTEM, print_header=False)
+
+
+def process_debugger():
+    if Options()['misc'].get('debug', False):
+        import pdb
+        pdb.post_mortem()
+
+
 def main(path_opts=None, run=None):
+    # first call to Options() load the options yaml file from --path_opts command line argument if path_opts=None
+    Options(path_opts)
+    # init options and exp dir for logging
+    init_experiment_directory(Options()['exp']['dir'], Options()['exp']['resume'])
```

Comment:
By the way, thanks a lot Micael :)
The debuging and profiling features are really needed!

Reply:
The standard behaviour of Options is definitely capable of compensating for the redundancy, therefore this is, indeed, possible.

On a side note, I don't see why anyone would want to write a custom `main` function (maybe for notebooks?), since exception management can be safely handled inside the engine; and in the super rare cases in which a very fine-grained exception management is required, I believe the user will have to be proficient enough to change the `run` function anyway — after all, these situations would probably also require substantial customization of bootstrap.

I will implement this for the sake of keeping clean compatibility with notebooks, but we might want to add this to the docs, otherwise we'll be the only two people in the world to know this is doable. ;)

Reply:
On a second pass, there are problems with this approach:
1. `init_experiments_directory` and `init_logs_options_files` are both _required_ for Logger to work. They have to be called before `activate_debugger` and `activate_profiler`.
2. `run` should not require an explicit construction call to Options. If this function is ever used inside a notebook, the user will have to call Options with `run_parser=False` prior to calling `run`, otherwise Options will try to argparse the command line arguments — which are non existant.

I'm pushing changes that include the Options' call, but in the notebook scenario the user will have to call Options beforehand, or things will crash. Furthermore, I'm duplicating the calls for `init_experiments_directory` and `init_logs_options_files` with an extra bypass parameter, both inside `run` and `main`, to accommodate for both scenarios. If `main` is called first, it will disable the calls inside `run`. This is necessary in order to keep `activate_debugger` and `activate_profiler` inside `main`, and at the same time allow `run` to be independent.